GMG/Scenes/ChordProgress/SegmentScrollView.swift

```diff
@@ -0,0 +1,106 @@
+    let currentChordCell: ChordCell?
+    let selectedChordCell: ChordCell?
+    let segmentHandlers: SegmentHandlers
+    let waveformHandlers: WaveformHandlers
+    let audioLevels: [Float]
+    let elapsedTime: TimeInterval
+    let isPlaying: Bool
+
+    @State private var isUserScrolling = false
+    @State private var isWaveformDragging = false
+
+    private var resolvedWaveformHandlers: WaveformHandlers {
+        WaveformHandlers(
+            onTap: waveformHandlers.onTap,
+            onDragStart: { time in
+                isWaveformDragging = true
+                waveformHandlers.onDragStart(time)
+            },
+            onDragChange: waveformHandlers.onDragChange,
+            onDragEnd: { time in
+                waveformHandlers.onDragEnd(time)
+                isWaveformDragging = false
+            }
+        )
+    }
+
+    private var shouldAutoScroll: Bool {
+        isPlaying && isUserScrolling == false && isWaveformDragging == false
+    }
+
+    private var currentSegmentIndex: Int? {
+        guard segmentSlices.isEmpty == false else { return nil }
+        let rawIndex = Int(elapsedTime / Constants.segmentDuration)
+        let clampedIndex = min(
+            max(0, rawIndex),
+            segmentSlices.count - 1
+        )
+        return clampedIndex
+    }
+
+    var body: some View {
+        ScrollViewReader { proxy in
+            ScrollView {
+                LazyVStack(spacing: Spacing.md) {
+                    ForEach(Array(segmentSlices.enumerated()), id: \.offset) { index, slices in
+                        Segment(
+                            index: index,
+                            totalDuration: totalDuration,
+                            chordSlices: slices,
+                            segmentDuration: Constants.segmentDuration,
+                            currentChordCell: currentChordCell,
+                            selectedChordCell: selectedChordCell,
+                            audioLevels: audioLevels,
+                            elapsedTime: elapsedTime,
+                            segmentHandlers: segmentHandlers,
+                            waveformHandlers: resolvedWaveformHandlers
+                        )
+                        .id(index)
+                    }
+                }
+                .safeAreaPadding(Spacing.md)
+                .safeAreaPadding(.bottom, 128)
+            }
+            .onScrollPhaseChange { _, newPhase in
+                isUserScrolling = newPhase == .interacting
+            }
+            .onChange(of: isPlaying) { _, isPlaying in
+                guard isPlaying, shouldAutoScroll else { return }
+                guard let index = currentSegmentIndex else { return }
+                withAnimation(.easeInOut(duration: 0.2)) {
+                    proxy.scrollTo(index, anchor: .center)
+                }
+            }
+            .onChange(of: currentSegmentIndex) { _, newIndex in
+                guard shouldAutoScroll else { return }
```

Comment:
```suggestion
               guard shouldAutoScroll, let newIndex = currentSegmentIndex else { return }
```

해당 부분도 마찬가지로요! 그리고 바로 아랫줄의 newIndex 검사하는 부분도 이렇게 한 줄로 만드는 것도 좋을 것 같아요!